src/v/iceberg/rest_client/catalog_client.h

```diff
@@ -0,0 +1,122 @@
+/*
+ * Copyright 2024 Redpanda Data, Inc.
+ *
+ * Licensed as a Redpanda Enterprise file under the Redpanda Community
+ * License (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * https://github.com/redpanda-data/redpanda/blob/master/licenses/rcl.md
+ */
+
+#pragma once
+
+#include "bytes/iobuf.h"
+#include "http/client.h"
+#include "http/request_builder.h"
+#include "iceberg/rest_client/retry_policy.h"
+#include "iceberg/rest_client/types.h"
+#include "json/document.h"
+#include "utils/named_type.h"
+#include "utils/retry_chain_node.h"
+
+#include <seastar/core/future.hh>
+#include <seastar/core/sstring.hh>
+
+namespace iceberg::rest_client {
+
+using base_path = named_type<ss::sstring, struct base_path_t>;
+using prefix_path = named_type<ss::sstring, struct prefix_t>;
+using api_version = named_type<ss::sstring, struct api_version_t>;
+
+// Holds parts of a root path used by catalog client
+struct path_components {
+    path_components(
+      std::optional<base_path> base = std::nullopt,
+      std::optional<prefix_path> prefix = std::nullopt,
+      std::optional<api_version> api_version = std::nullopt);
+
+    // Returns root path for use in API calls not related to oauth token, joins
+    // base, api version and prefix with slashes and adds a terminating slash.
+    ss::sstring root_path() const;
+    // Same as root path but does not include the prefix
+    ss::sstring token_api_path() const;
```

Comment:
added a couple of examples from polaris, also I realized that token path can be simpler, since the spec fixes it to `oauth/tokens`, the client does not need to compute it, it can be computed in `path_components`